### PR TITLE
Add OSSL_CMP_CTX_get1_newChain() and related CLI option -chainout

### PR DESCRIPTION
**Pull request by DDvO**
This adds the CMP CLI option `-chainout` and simplifies certificate saving in `apps/cmp.c`.

### CODE REVIEW

**Comment by DDvO** [comment]
This PR carves out a rather independent piece of the larger PR #12741 for simpler review.

Could you please review this swiftly such that this API extension can be merged before feature freeze.

File: apps/cmp.c [hunk 1]
```diff
@@ -2935,39 +2945,26 @@ int cmp_main(int argc, char **argv)
...
+            if (!X509_add_cert(certs, newcert, X509_ADD_FLAG_UP_REF)) {
                 sk_X509_free(certs);
                 goto err;
             }
-            sk_X509_free(certs);
+            if (save_free_certs(cmp_ctx, certs, opt_certout, "enrolled") < 0)
+                goto err;
         }
+        if (save_free_certs(cmp_ctx, OSSL_CMP_CTX_get1_newChain(cmp_ctx),
+                            opt_chainout, "extra") < 0)
```
Comment (mattcaswell) [inline apps/cmp.c:2965]: Is the description here correct?

File: crypto/cmp/cmp_ctx.c [hunk 2]
```diff
@@ -459,6 +461,34 @@ int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
...
+    if (ctx->newChain == NULL)
+        return sk_X509_new_null();
+    return X509_chain_up_ref(ctx->newChain);
+}
+
+/*
+ * Copies any given stack of inbound X509 certificates to newChain
+ * of the OSSL_CMP_CTX structure so that they may be retrieved later.
+ */
+int ossl_cmp_ctx_set1_newChain(OSSL_CMP_CTX *ctx, STACK_OF(X509) *newChain)
```
Comment (mattcaswell) [inline crypto/cmp/cmp_ctx.c:480]: I'm failing to see where this gets called with anything other than NULL for its last param?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: apps/cmp.c (hunk 1 above)
Comment (DDvO) [reply apps/cmp.c]: Oops, this was a copy&paste glitch - fixed.

File: crypto/cmp/cmp_ctx.c (hunk 2 above)
Comment (DDvO) [reply crypto/cmp/cmp_ctx.c:480]: Right, in this PR it is not really used, and this is an artifact of my mentioned recent split of PRs.
It is going to be used after merging in #12741, which currently contains in `crypto/cmp/cmp_client.c:`
```
    /* TODO(3.0) add when available: (void)ossl_cmp_ctx_set1_newChain(ctx, chain); */   
```

**Comment by DDvO** [comment]
Thanks @mattcaswell for your two review comments - I've handled them.

**Review by mattcaswell — APPROVED** [review]

**Comment by DDvO** [comment]
This was quick :)
Concurrently I've found a second copy&paste glitch and corrected it: in a comment in `cmp_local.h`:
```
-    STACK_OF(X509) *newChain; /* CA certs received from server (in IP message) */
+    STACK_OF(X509) *newChain; /* chain of newly enrolled cert received */
```
Could you please re-approve (as far as needed for that trivial change)?

**Review by mattcaswell — APPROVED** [review]
Reconfirmed

**Comment by DDvO** [comment]
Squashed and merged - thanks

**Pull request closed without merging**